### ISSUE TITLE
[Parchment] Unable to create mention blot

### ISSUE THREAD

**Issue by essivision** [afconsult/quill-mention#76]
I use this module in Quill but error me:
```
[Parchment] Unable to create mention blot
```

I used this:
```
 import mention from 'quill-mention';
Quill.register({
     'modules/mention': mention,
});
```

**Comment by jurgen-reconcept** [comment]
I think this has something to do with Quill v2. I had the same problem after updating Quill from 1.x to 2.0.0-dev.3.

What fixed it for me was pulling this whole repo and making a local version.

**Comment by mtimofiiv** [comment]
This problem is still around. I figured out what goes wrong in my case...

The `quill-mention` lib tries to import `Quill` via `require('quill)`/`import Quill from 'quill'`.

But if you're using `import Quill from 'quill/dist/quill.core'` in the implementation, that's where it fails.

So this issue is actually rather similar to https://github.com/afry/quill-mention/issues/77

My quick and dirty fix is essentially to do a core-only build of this package, but that's obviously not ideal: https://github.com/parlay-work/quill-mention

I would like to offer a suggestion on how to fix this – perhaps don't self-register with `Quill` on import and let the developer run `Quill.register` themselves?

**Comment by lithxe** [comment]
Could I get more insight on how to fix this? I am still having this problem with Quill 2 and QuillMention. :(

**Comment by NakedFool** [comment]
> 
> 
> This problem is still around. I figured out what goes wrong in my case...
> 
> The `quill-mention` lib tries to import `Quill` via `import Quill from "quill"` (https://github.com/afry/quill-mention/blob/master/src/quill.mention.js#L1)
> 
> But if you're using `import Quill from 'quill/dist/quill.core'` in your app, that's where it fails.
> 
> So this issue is actually rather similar to #77
> 
> My quick and dirty fix is essentially to do a core-only build of this package, but that's obviously not ideal.
> 
> I would like to offer a suggestion on how to fix this – perhaps don't self-register with `Quill` on import and let the developer run `Quill.register` themselves?
> 
> You could potentially export the blot and the module separately and allow the developer to do something like:
> 
> ```js
> import Quill from 'quill'
> import { QuillMentionBlot, QuillMentionModule } from 'quill-mention'
> 
> Quill.register(QuillMentionBlot)
> Quill.register('modules/mention', QuillMentionModule)
> ```
> 
> This would be a breaking API change for the package, but something to consider. Also possible to export a constructing function that takes `Quill` as an argument:
> 
> ```js
> import Quill from 'quill'
> import QuillMention from 'quill-mention'
> 
> QuillMention(Quill)
> ```

Hey may i ask you how you solved this? As when i try with your first approach where i import Blot and Module from 'quill-mention' it says that they can't be found.

**Comment by mtimofiiv** [comment]
@NakedFool I actually ended up doing what @jurgen-reconcept suggests in that thread and clone the package, run `npm run build`, copy the files in `./dist` into your own project and use that. This way, you force the plugin to use the version of Quill you're using, in your case 2.0.0.

TLDR; explanation:

See, when you create a plugin you need to register it, by running `Quill.register(MyPlugin)`. This is [well documented](https://quilljs.com/guides/building-a-custom-module/) and all but a lot of plugins (like the Emoji one as well) run `Quill.register()` for you inside their own module.

They do this cause then there's less boilerplate code for devs wanting to use their plugins, but the downside is that they need to import Quill themselves into their module in order to run its static method. So in this module's dependencies you'll find `quill@1.3.x` and `npm` resolves this as its own version of Quill, meaning your code runs `quill@2.0.0` and this plugin has a different version, `quill@1.3.7`. Here's where this goes wrong.

The problem I was having was actually not with 2.0.0 as much as using the `quill/core` version of Quill. the solution is very similar. Clone the repo, go into `src/quill.mention.js` and `src/blots/mention.js` and edit the `import Quill from 'quill'` to `import Quill from 'quill/core'`. Do your own build. Copy files from `./dist` and viola.

**Comment by NakedFool** [comment]
> 
> 
> @NakedFool I actually ended up doing what @jurgen-reconcept suggests in that thread and clone the package, run `npm run build`, copy the files in `./dist` into your own project and use that. This way, you force the plugin to use the version of Quill you're using, in your case 2.0.0.
> 
> TLDR; explanation:
> 
> See, when you create a plugin you need to register it, by running `Quill.register(MyPlugin)`. This is [well documented](https://quilljs.com/guides/building-a-custom-module/) and all but a lot of plugins (like the Emoji one as well) run `Quill.register()` for you inside their own module.
> 
> They do this cause then there's less boilerplate code for devs wanting to use their plugins, but the downside is that they need to import Quill themselves into their module in order to run its static method. So in this module's dependencies you'll find `quill@1.3.x` and `npm` resolves this as its own version of Quill, meaning your code runs `quill@2.0.0` and this plugin has a different version, `quill@1.3.7`. Here's where this goes wrong.
> 
> The problem I was having was actually not with 2.0.0 as much as using the `quill/core` version of Quill. the solution is very similar. Clone the repo, go into `src/quill.mention.js` and `src/blots/mention.js` and edit the `import Quill from 'quill'` to `import Quill from 'quill/core'`. Do your own build. Copy files from `./dist` and viola.

Thank you man, thanks @jurgen-reconcept for coming to solution and you for your concise explanation.
So when it comes to this exact error message like title of the issue local version of 'quill-mention' solves problem as code runs your actually version of quill instead of one defined by package when installed.

**Comment by DevJoaoLopes** [comment]
I still have this problem, any consistent solution?

**Comment by DevJoaoLopes** [comment]
@NakedFool is there any other way to solve this problem other than via local?

**Comment by mtimofiiv** [comment]
@DevJoaoLopes as long as this module specifies the dependency [`quill@^1.3.4` in `package.json`](https://github.com/afry/quill-mention/blob/master/package.json#L16) and [runs `Quill.register()`](https://github.com/afry/quill-mention/blob/master/src/quill.mention.js#L741) on this self-specified copy of Quill instead of whatever one you're using in your code, you're out of luck.

Try forking it and building it yourself if you want something a bit more sustainable? Unless one of the maintainers weighs in on this.

**Comment by hasangursoy** [comment]
If you're setting formats, make sure you have 'mention' in that list.
